File: src/runtime.c
#include <stdio.h>

void colang_print_int(int val) {
    printf("%d", val);
    fflush(stdout);
}

void colang_print_str(const char* str) {
    if (str) {
        printf("%s", str);
        fflush(stdout);
    }
}

void colang_input_int(int* var_ptr) {
    if (var_ptr) {
        if (scanf("%d", var_ptr) != 1) {
            *var_ptr = 0;
        }
    }
}

File: srcpas/runtime.pas
unit runtime;

{$mode objfpc}{$H+}

interface

uses
  SysUtils;

// Экспортируемые процедуры для обращения из NASM (соглашение cdecl)
procedure colang_print_int(Val: LongInt); cdecl; export;
procedure colang_print_str(StrPtr: PChar); cdecl; export;
procedure colang_input_int(VarPtr: PLongInt); cdecl; export;

implementation

procedure colang_print_int(Val: LongInt); cdecl;
begin
  Write(Val);
  Flush(Output);
end;

procedure colang_print_str(StrPtr: PChar); cdecl;
begin
  if StrPtr <> nil then
  begin
    Write(StrPtr);
    Flush(Output);
  end;
end;

procedure colang_input_int(VarPtr: PLongInt); cdecl;
var
  InputVal: LongInt;
begin
  if VarPtr <> nil then
  begin
    Read(InputVal);
    VarPtr^ := InputVal;
  end;
end;

end.
